compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,132 +75,280 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
+/// A DmaBatch contains the following :-
+///   1. DmaOps within the current batch.
+///   2. Required Bd Ids by the current batch.
+///   3. Nested DmaBatch chain.
+///   4. Pointer to the sibling DmaBatch.
+///   4. A pointer to the parent op if the parent of is a scf.for.
+typedef struct DmaBatchStruct {
+  SmallVector<AMDAIE::NpuDmaCpyNdOp> currentDmaOps = {};
+  int32_t requiredBdIds = 0;
+  SmallVector<DmaBatchStruct *> immediateInnerBatches = {};
+  DmaBatchStruct *nextDmaBatch = nullptr;
+  scf::ForOp forOpParent = nullptr;
+} DmaBatch;
+
+using TileDmaBatchGraph = llvm::MapVector<AMDAIE::TileOp, DmaBatch *>;
+
+/// Create a new TileDmaBatchGraph by traversing over each tiles in a workgroup.
+static TileDmaBatchGraph initTileDmaBatchGraph(
+    AMDAIE::WorkgroupOp workgroupOp) {
+  TileDmaBatchGraph tileDmaBatchGraph;
+  workgroupOp.walk([&](AMDAIE::TileOp tile) {
+    if (tileDmaBatchGraph.contains(tile)) return WalkResult::skip();
+    tileDmaBatchGraph[tile] = new DmaBatch();
+    return WalkResult::advance();
+  });
+  return tileDmaBatchGraph;
+}
+
+/// A DmaBatch with no DmaOps and no nested DmaBatches is an empty DmaBatch.
+static bool isEmptyDmaBatch(DmaBatch *dmaBatch) {
+  return (dmaBatch->currentDmaOps.empty() &&
+          dmaBatch->immediateInnerBatches.empty());
+}
+
+static void updateInnerBatchesOfCurrentTileDmaBatchGraph(
+    DenseMap<AMDAIE::TileOp, DmaBatch *> &currentTileBatch,
+    TileDmaBatchGraph &innerTileDmaBatchGraph) {
+  for (auto [tile, dmaBatch] : innerTileDmaBatchGraph) {
+    if (isEmptyDmaBatch(dmaBatch)) continue;
+    currentTileBatch[tile]->immediateInnerBatches.push_back(dmaBatch);
+  }
+}
+
+static TileDmaBatchGraph formTileDmaBatchGraph(
+    AMDAIE::WorkgroupOp &workgroupOp, Operation *parentOp,
+    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap) {
+  TileDmaBatchGraph tileDmaBatchGraph = initTileDmaBatchGraph(workgroupOp);
+
+  DenseMap<AMDAIE::TileOp, DmaBatch *> currentTileBatch;
+  for (auto [tile, dmaBatch] : tileDmaBatchGraph) {
+    currentTileBatch[tile] = dmaBatch;
+    currentTileBatch[tile]->forOpParent = dyn_cast<scf::ForOp>(parentOp);
+  }
+  auto addDmaToBatch = [&](AMDAIE::TileOp tile, AMDAIE::NpuDmaCpyNdOp dmaOp) {
+    assert(tileDmaBatchGraph.contains(tile) && "Tile op not found");
+    currentTileBatch[tile]->currentDmaOps.push_back(dmaOp);
+  };
+
+  auto updateCurrentTileBatch = [&](AMDAIE::TileOp tile) {
+    currentTileBatch[tile]->nextDmaBatch = new DmaBatch();
+    currentTileBatch[tile] = currentTileBatch[tile]->nextDmaBatch;
+    currentTileBatch[tile]->forOpParent = dyn_cast<scf::ForOp>(parentOp);
+  };
+
+  auto updateInnerBatchesOfCurrentTileDmaBatchGraph =
+      [&](TileDmaBatchGraph &innerTileDmaBatchGraph) {
+        for (auto [tile, dmaBatch] : innerTileDmaBatchGraph) {
+          if (isEmptyDmaBatch(dmaBatch)) continue;
+          currentTileBatch[tile]->immediateInnerBatches.push_back(dmaBatch);
+        }
+      };
```

Comment:
It looks useful to create a class/struct to maintain `currentTileBatch` and add these methods to the class instead.